lib/trino-parquet/src/main/java/io/trino/parquet/writer/valuewriter/Int96TimestampValueWriter.java

```diff
@@ -0,0 +1,108 @@
+import static io.trino.spi.type.Timestamps.NANOSECONDS_PER_MILLISECOND;
+import static io.trino.spi.type.Timestamps.PICOSECONDS_PER_NANOSECOND;
+import static java.lang.Math.floorDiv;
+import static java.lang.Math.floorMod;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class Int96TimestampValueWriter
+        extends PrimitiveValueWriter
+{
+    private final TimestampType timestampType;
+    private final DateTimeZone timeZone;
+
+    public Int96TimestampValueWriter(ValuesWriter valuesWriter, Type type, PrimitiveType parquetType, DateTimeZone timeZone)
+    {
+        super(parquetType, valuesWriter);
+        this.timestampType = (TimestampType) requireNonNull(type, "type is null");
+        checkArgument(
+                parquetType.getPrimitiveTypeName().equals(PrimitiveType.PrimitiveTypeName.INT96),
+                "parquetType %s is not INT96");
+        this.timeZone = requireNonNull(timeZone, "timeZone is null");
+    }
+
+    @Override
+    public void write(Block block)
+    {
+        int positionCount = block.getPositionCount();
+        long[] localEpochMillis = new long[positionCount];
+        int[] nanosOfMillis = new int[positionCount];
+        int nonNullsCount = 0;
+        if (timestampType.isShort()) {
+            for (int position = 0; position < positionCount; position++) {
+                if (!block.isNull(position)) {
+                    long epochMicros = timestampType.getLong(block, position);
+                    localEpochMillis[nonNullsCount] = floorDiv(epochMicros, MICROSECONDS_PER_MILLISECOND);
+                    nanosOfMillis[nonNullsCount] = floorMod(epochMicros, MICROSECONDS_PER_MILLISECOND) * NANOSECONDS_PER_MICROSECOND;
+                    nonNullsCount++;
+                }
+            }
+        }
+        else {
+            for (int position = 0; position < positionCount; position++) {
+                if (!block.isNull(position)) {
+                    LongTimestamp timestamp = (LongTimestamp) timestampType.getObject(block, position);
+                    long epochMicros = timestamp.getEpochMicros();
+                    localEpochMillis[nonNullsCount] = floorDiv(epochMicros, MICROSECONDS_PER_MILLISECOND);
+                    int nanosOfMicro = timestamp.getPicosOfMicro() / PICOSECONDS_PER_NANOSECOND;
+                    nanosOfMillis[nonNullsCount] = (floorMod(epochMicros, MICROSECONDS_PER_MILLISECOND) * NANOSECONDS_PER_MICROSECOND) + nanosOfMicro;
+                    nonNullsCount++;
+                }
+            }
+        }
+
+        for (int i = 0; i < nonNullsCount; i++) {
+            long epochMillis = timeZone.convertLocalToUTC(localEpochMillis[i], false);
+            long epochDay = floorDiv(epochMillis, MILLISECONDS_PER_DAY);
+            int julianDay = JULIAN_EPOCH_OFFSET_DAYS + toIntExact(epochDay);
+
+            long nanosOfEpochDay = (epochMillis - (epochDay * MILLISECONDS_PER_DAY)) * NANOSECONDS_PER_MILLISECOND;
```

Comment:
`(epochMillis - (epochDay * MILLISECONDS_PER_DAY))` -> 
`floorMod(epochMillis, MILLISECONDS_PER_DAY)`

(use `floorDiv` and `floorMod` in tandem consistently)

Reply:
I had avoided that because IDE starts to complain about `integer multiplication implicitly cast to long`.
I'll update it to `(long) floorMod(epochMillis, MILLISECONDS_PER_DAY)` since that seems easier to read.